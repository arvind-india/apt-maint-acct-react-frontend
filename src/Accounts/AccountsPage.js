import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { Table } from 'reactstrap'
import { Router } from 'react-router-dom'

import {
  MdAdd,
  MdVisibility,
  MdEdit,
  MdDelete
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

import { history } from '../_helpers'
import { PrivateRoute, FlashMessage } from '../_components'
import { accountActions as actions, userActions } from '../_actions'
import { AccountDetailsPage as detailsPage } from './AccountDetailsPage'

let url = '/accounts'
let module = 'accounts'

class AccountsPage extends React.Component {

  constructor(props) {
    super(props)
    let noOfPrevMonths = 2  // by default show all records in the past 2 months
    let year = new Date().getFullYear();
    year = 2016 // for testing purpose only
    let month = new Date().getMonth() - noOfPrevMonths;
    this.state = {
      modal: false,
      canDelete: false,
      fromDate: new Date(year, month, 1),
      toDate: new Date()
    }
    this.handleDeleteModel = this.handleDeleteModel.bind(this)
//    this.toggleModalDialog = this.toggleModalDialog.bind(this)
//    this.deleteConfirmed = this.deleteConfirmed.bind(this)
  }

  componentDidMount() {
    const { fromDate, toDate } = this.state
    //this.props.dispatch(actions.getAll())
    this.props.dispatch(actions.getListFor(fromDate, toDate))
    this.props.dispatch(userActions.getAll())
  }

  render() {
    console.log('Props in AccountsPage: ', this.props)
    const { accounts, alert, authzn, users } = this.props
    let models = accounts
console.log('accounts models: ', models)
    return (
      <div>
        <h3>Accounts List</h3>
        {alert.message && <FlashMessage text={alert.message} delay={5000}/>}
        {models.items && authzn && users.items && this.showList() }
        <Router history={history}>
          <div>
            <PrivateRoute path={`${url}/:id`} component={detailsPage} />
          </div>
        </Router>
      </div>
    )
  }

  showList(){
    const { accounts } = this.props
    let models = accounts
    return <Table>
      <thead>{ this.headerRow() }</thead>
      <tbody>
        { models.items.map((model, index) => this.bodyRow(model,index)) }
      </tbody>
    </Table>
  }

  headerRow() {
    return <tr>
      <th>#</th>
      <th>Txn Date</th>
      <th>Flat#</th>
      <th>Name</th>
      <th>Mth</th>
      <th>Yr</th>
      <th>Cr/Dr</th>
      <th>Amt in &#8377;</th>
      <th>&#8377; Balance</th>
      <th>Category</th>
      <th>Actions {this.addLink()}</th>
    </tr>
  }
  addLink() {
    const { authzn } = this.props
    return authzn.allowsAdd ?
      <Link to={{ pathname: `${url}/0`, state: this.newModel() }} title="Add"><MdAdd/></Link> :
      ''
  }
  newModel() {
    return {
      model: {
        id: 0,
        recorded_at: '',
        item: '',
        flat_number:'',
        name: '',
        for_month: '',
        for_year: '',
        crdr: '',
        amount: '',
        balance: '',
        category: '',
        remarks: ''
      }
    }
  }

  bodyRow(model,index) {
    return <tr key={model.id}>
      <td>{index+1}</td>
      <td>{model.recorded_at}</td>
      <td>{model.flat_number}</td>
      <td>{model.name}</td>
      <td>{model.for_month}</td>
      <td>{model.for_year}</td>
      <td>{model.crdr}</td>
      <td>{model.amount}</td>
      <td>{model.balance}</td>
      <td>{model.category}</td>
      {this.showActions(model)}
    </tr>
  }

  showActions(model) {
    const { authzn } = this.props
    return <td>
            <Link
              to={{ pathname: `${url}/${model.id}`, state:{model: model} }}
              title={authzn.allowsEdit?"Edit":"View"}
            >{authzn.allowsEdit?<MdEdit/>:<MdVisibility/>}</Link>
            <Button
              color="link"
              title="Delete"
              onClick={() => this.handleDeleteModel(model.id)}
              hidden={!authzn.allowsDelete}
            ><MdDelete color="red"/></Button>
          </td>
  }

  handleDeleteModel(id) {
    const { fromDate, toDate } = this.state
    console.log('Deleting Account with id: ', id)

    if( window.confirm('Are you sure?') ) {
      console.log('Delete confirmed')
      this.props.dispatch(actions.delete(id))
      this.props.dispatch(actions.getListFor(fromDate, toDate))
    } else {
      console.log('Delete cancelled!')
    }
  }

}

function mapStateToProps(state) {
  const { accounts, alert, authorizations, users } = state
  // const { user } = authentication
  const authzn = authorizations[module]
  return {
//    user,
    accounts,
    alert,
    authzn,
    users
  }
}

const connectedAccountsPage = connect(mapStateToProps)(AccountsPage)
export { connectedAccountsPage as AccountsPage }
