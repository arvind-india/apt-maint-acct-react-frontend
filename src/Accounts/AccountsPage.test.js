// AccountsPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { Accounts } from './AccountsPage'

const mockProps = {
  trackHistory: false,  // for unit testing history is NOT required as it has 'keys' that are dynamic
  accounts: {
    loading: false,
    error: false,
    items: [
      {
        id: 1,
        recorded_at: '2018-04-04',
        item: 'maintenance fee',
        flat_number: 'G1',
        name: 'tester',
        for_month: '04',
        for_year: '2018',
        crdr: 'cr',
        amount: '600',
        balance: '',
        category: 'maintenance',
        remarks: 'maintenance for the month april 2018'
      },
      {
        id: 2,
        recorded_at: '2018-04-01',
        item: 'sewage clearance',
        flat_number: 'NA',
        name: 'Sewage Truck Service',
        for_month: '04',
        for_year: '2018',
        crdr: 'dr',
        amount: '2200',
        balance: '',
        category: 'sewage disposal',
        remarks: 'septic tank clearance'      },
      {
        id: 3,
        recorded_at: '2018-03-31',
        item: 'Garbage clearance',
        flat_number: 'NA',
        name: 'Panchayat Garbage disposal Service',
        for_month: '04',
        for_year: '2018',
        crdr: 'dr',
        amount: '200',
        balance: '',
        category: 'maintenance',
        remarks: 'Garbage clearance'
      }
    ]
  },
  users: {
    items: []
  },
  alert: {message: "testingAlert"},
  authzn: {
    allowsAdd: true,
    allowsEdit: true,
    allowsDelete: true,
    condition: ''
  },
  getAll: jest.fn(),
  delete: jest.fn(),
  getListFor: jest.fn(),
  setSessionStorage: jest.fn()
}

const component = shallow(<Accounts {...mockProps} />)

describe('AccountsPage params test', () => {
  it('should render AccountsPage without crashing', () => {
    component
  })

  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
})

describe('AccountsPage rows test', () => {
  let numberOfMockAccounts = mockProps.accounts.items.length
  let tbody = component.find('tbody')
  let rows = tbody.find('tr')
  it('should have '+numberOfMockAccounts+' rows', () => {
    expect(rows.length).toEqual(numberOfMockAccounts)
  })
})
