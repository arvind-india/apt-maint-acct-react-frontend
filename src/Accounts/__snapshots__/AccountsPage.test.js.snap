// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Accounts List
    </h3>
    <FlashMessage
        color={undefined}
        delay={3000}
        text="testingAlert"
    />
    <div
        className="grid-form"
    >
        <div
            data-row-span="6"
        >
            <div
                data-field-span="1"
            >
                <Label
                    tag="label"
                    widths={
                        Array [
                            "xs",
                            "sm",
                            "md",
                            "lg",
                            "xl",
                          ]
                    }
                >
                    From
                </Label>
                <Input
                    name="fromDate"
                    onChange={[Function]}
                    placeholder="Accounts from this date"
                    tag="p"
                    type="date"
                    value="2018-02-06"
                />
            </div>
            <div
                data-field-span="1"
            >
                <Label
                    tag="label"
                    widths={
                        Array [
                            "xs",
                            "sm",
                            "md",
                            "lg",
                            "xl",
                          ]
                    }
                >
                    To
                </Label>
                <Input
                    name="toDate"
                    onChange={[Function]}
                    placeholder="Accounts from this date"
                    tag="p"
                    type="date"
                    value="2018-04-06"
                />
            </div>
        </div>
    </div>
    <Table
        responsiveTag="div"
        tag="table"
    >
        <thead>
            <tr>
                <th>
                    #
                </th>
                <th>
                    Txn Date
                </th>
                <th>
                    Flat#
                </th>
                <th>
                    Name
                </th>
                <th>
                    Mth
                </th>
                <th>
                    Yr
                </th>
                <th>
                    Cr/Dr
                </th>
                <th>
                    Amt in ₹
                </th>
                <th>
                    ₹ Balance
                </th>
                <th>
                    Category
                </th>
                <th>
                    Actions 
                    <Link
                        replace={false}
                        title="Add"
                        to={
                            Object {
                                "pathname": "/accounts/0",
                                "state": Object {
                                  "model": Object {
                                    "amount": "",
                                    "balance": "",
                                    "category": "",
                                    "crdr": "",
                                    "flat_number": "",
                                    "for_month": "",
                                    "for_year": "",
                                    "id": 0,
                                    "item": "",
                                    "name": "",
                                    "recorded_at": "",
                                    "remarks": "",
                                  },
                                },
                              }
                        }
                    >
                        <MdAdd />
                    </Link>
                </th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>
                    1
                </td>
                <td>
                    2018-04-04
                </td>
                <td>
                    G1
                </td>
                <td>
                    tester
                </td>
                <td>
                    04
                </td>
                <td>
                    2018
                </td>
                <td>
                    cr
                </td>
                <td>
                    600
                </td>
                <td />
                <td>
                    maintenance
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/accounts/1",
                                "state": Object {
                                  "model": Object {
                                    "amount": "600",
                                    "balance": "",
                                    "category": "maintenance",
                                    "crdr": "cr",
                                    "flat_number": "G1",
                                    "for_month": "04",
                                    "for_year": "2018",
                                    "id": 1,
                                    "item": "maintenance fee",
                                    "name": "tester",
                                    "recorded_at": "2018-04-04",
                                    "remarks": "maintenance for the month april 2018",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <td>
                    2
                </td>
                <td>
                    2018-04-01
                </td>
                <td>
                    NA
                </td>
                <td>
                    Sewage Truck Service
                </td>
                <td>
                    04
                </td>
                <td>
                    2018
                </td>
                <td>
                    dr
                </td>
                <td>
                    2200
                </td>
                <td />
                <td>
                    sewage disposal
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/accounts/2",
                                "state": Object {
                                  "model": Object {
                                    "amount": "2200",
                                    "balance": "",
                                    "category": "sewage disposal",
                                    "crdr": "dr",
                                    "flat_number": "NA",
                                    "for_month": "04",
                                    "for_year": "2018",
                                    "id": 2,
                                    "item": "sewage clearance",
                                    "name": "Sewage Truck Service",
                                    "recorded_at": "2018-04-01",
                                    "remarks": "septic tank clearance",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <td>
                    3
                </td>
                <td>
                    2018-03-31
                </td>
                <td>
                    NA
                </td>
                <td>
                    Panchayat Garbage disposal Service
                </td>
                <td>
                    04
                </td>
                <td>
                    2018
                </td>
                <td>
                    dr
                </td>
                <td>
                    200
                </td>
                <td />
                <td>
                    maintenance
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/accounts/3",
                                "state": Object {
                                  "model": Object {
                                    "amount": "200",
                                    "balance": "",
                                    "category": "maintenance",
                                    "crdr": "dr",
                                    "flat_number": "NA",
                                    "for_month": "04",
                                    "for_year": "2018",
                                    "id": 3,
                                    "item": "Garbage clearance",
                                    "name": "Panchayat Garbage disposal Service",
                                    "recorded_at": "2018-03-31",
                                    "remarks": "Garbage clearance",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
        </tbody>
    </Table>
    <Router
        history={Object {}}
    >
        <div>
            <PrivateRoute
                component={[Function]}
                path="/accounts/:id"
            />
        </div>
    </Router>
</div>,
]
`;
